src/theory/bags/normal_form.cpp

```diff
@@ -12,26 +12,628 @@
+    return n;
+  }
+  switch (n.getKind())
+  {
+    case MK_BAG: return evaluateMakeBag(n);
+    case BAG_COUNT: return evaluateBagCount(n);
+    case UNION_DISJOINT: return evaluateUnionDisjoint(n);
+    case UNION_MAX: return evaluateUnionMax(n);
+    case INTERSECTION_MIN: return evaluateIntersectionMin(n);
+    case DIFFERENCE_SUBTRACT: return evaluateDifferenceSubtract(n);
+    case DIFFERENCE_REMOVE: return evaluateDifferenceRemove(n);
+    case BAG_CHOOSE: return evaluateChoose(n);
+    case BAG_CARD: return evaluateCard(n);
+    case BAG_IS_SINGLETON: return evaluateIsSingleton(n);
+    case BAG_FROM_SET: return evaluateFromSet(n);
+    case BAG_TO_SET: return evaluateToSet(n);
+    default:
+    {
+    }
+    break;
+  }
+  Unhandled() << "Unexpected bag kind '" << n.getKind() << "' in node " << n
+              << std::endl;
+}
+
+template <typename T1, typename T2, typename T3, typename T4, typename T5>
+Node NormalForm::evaluateBinaryOperation(const TNode& n,
+                                         T1&& equal,
+                                         T2&& less,
+                                         T3&& greaterOrEqual,
+                                         T4&& remainderOfA,
+                                         T5&& remainderOfB)
+{
+  std::map<Node, Rational> elementsA = getBagElements(n[0]);
+  std::map<Node, Rational> elementsB = getBagElements(n[1]);
+  std::map<Node, Rational> elements;
+
+  std::map<Node, Rational>::const_iterator itA = elementsA.begin();
+  std::map<Node, Rational>::const_iterator itB = elementsB.begin();
+
+  Trace("bags-evaluate") << "[NormalForm::evaluateBinaryOperation "
+                         << n.getKind() << "] " << std::endl
+                         << "elements A: " << elementsA << std::endl
+                         << "elements B: " << elementsB << std::endl;
+
+  while (itA != elementsA.end() && itB != elementsB.end())
+  {
+    if (itA->first == itB->first)
+    {
+      equal(elements, itA, itB);
+      itA++;
+      itB++;
+    }
+    else if (itA->first < itB->first)
+    {
+      less(elements, itA, itB);
+      itA++;
+    }
+    else
+    {
+      greaterOrEqual(elements, itA, itB);
+      itB++;
+    }
+  }
+
+  // handle the remaining elements from A
+  remainderOfA(elements, elementsA, itA);
+  // handle the remaining elements from B
+  remainderOfA(elements, elementsB, itB);
+
+  Trace("bags-evaluate") << "elements: " << elements << std::endl;
+  Node bag = constructBagFromElements(n.getType(), elements);
+  Trace("bags-evaluate") << "bag: " << bag << std::endl;
+  return bag;
+}
+
+std::map<Node, Rational> NormalForm::getBagElements(TNode n)
+{
+  Assert(n.isConst()) << "node " << n << " is not in a normal form"
+                      << std::endl;
+  std::map<Node, Rational> elements;
+  if (n.getKind() == EMPTYBAG)
+  {
+    return elements;
+  }
+  while (n.getKind() == kind::UNION_DISJOINT)
+  {
+    Assert(n[0].getKind() == kind::MK_BAG);
```

Comment:
https://github.com/CVC4/CVC4/pull/5294#discussion_r509273946
Perhaps this assertion is not necessary. `n` is asserted to be constant in line 169